loader/0032-loader-plugins.md

```diff
@@ -0,0 +1,328 @@
+and install APIs and libraries in addition to the mods they want to use. Adding a
+way for Quilt Loader to fetch dependencies dynamically will help ease this
+headache and make Quilt a more appealing option for new mod users. Plugins will
+allow us to add such behavior dynamically when required, while allowing the
+behavior to be disabled if a user does not want or need it.
+
+Plugins will allow third-party libraries to define complex mod resolution behavior
+such as [allowing mods to declare "API" classes that other mods can declare implementations for](https://github.com/FabricMC/fabric-loader/issues/343).
+
+Projects such as [PatchworkMC](https://patchworkmc.net/) and [Sponge](https://www.spongepowered.org/) will be able to run non-Quilt mods ([Forge](https://forums.minecraftforge.net/) and Sponge mods, respectively) on top of the Quilt platform by
+performing transformations during the mod loading process.
+
+Once CHASM is implemented as a universal standard for bytecode transformations
+across Quilt, this system will allow us to remove the special case code for Mixin
+and access wideners and reimplement those as loader plugins instead. This will make
+developing third-party libraries that modify bytecode in unique and interesting
+ways possible.
+
+Loader plugins could also be used to [add mod containers for mods that don't do anything](https://github.com/FabricMC/fabric-loader/issues/175), although a
+better system for that will likely be implemented separately into Loader itself.
+
+A loader plugin could be developed to load mods from instance-specific folders instead of the single shared mods folder.
+
+By breaking parts of the mod loading process into various projects, we can keep
+each repository focused on its own specific mission. This will allow for more
+finely grained separation of responsibility.
+
+
+# Explanation
+
+The addition of Quilt Loader plugins will make Quilt Loader a plugin loader
+first, and mod loader second. Much of the existing mod loading behavior will be
+moved to one or more first-party loader plugins that are installed by default,
+keeping only the bare minimum amount of code required for mod resolving in Loader
+proper. Loader plugins will be capable of effecting the mod loading process by
+adding/removing mods and even adding other plugins. Plugins will also have the
+ability to open/create UI elements which can be used for as simple a thing as
+prompting the user for confirmation, or as complex of one as writing a full-scale
+mod launcher as a loader plugin.
+
+## Application
+Loader plugins are initially added from the classpath, allowing default loader
+plugins for an instance to be added via the profile json. All plugins are then
+applied in cycles. During each cycle, the following things happen:
+- Each plugin is given a chance to add any plugin candidates
+    - Each new plugin added this way is given a chance to add any plugin candidates.
+      Rinse and repeat.
```

Comment:
Can all plugins see (in classpath terms) all plugins which have loaded up to the point they run? Can they talk between each other directly? Suppose this is linked to the library plugin question